R/stop_paste_linter.R

```diff
@@ -0,0 +1,52 @@
+#' Block usage of paste() and paste0() with messaging functions using ...
+#'
+#' `stop(paste0(...))` is strictly redundant -- `stop(...)` is equivalent.
+#'   `stop(...)` is also preferable to `stop(paste(...))`.
+#'
+#' @evalRd rd_tags("stop_paste_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+stop_paste_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$xml_parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    translators <- c("packageStartupMessage", "message", "warning", "stop")
+    xpath <- glue::glue("//expr[
+      expr[SYMBOL_FUNCTION_CALL[ {xp_text_in_table(translators)} ]]
+      and expr[
+        expr[SYMBOL_FUNCTION_CALL[text() = 'paste' or text() = 'paste0']]
+        and not(SYMBOL_SUB[
+          text() = 'collapse'
+          or (
+            text() = 'sep'
+            and not(following-sibling::expr[1][STR_CONST[text() = '\"\"' or text() = '\" \"']])
```

Comment:
This won't capture all string literals representing the empty string or a space.

Reply:
left as a TODO for now